k2/csrc/fsa_algo.cc

```diff
@@ -44,27 +48,62 @@ inline int32_t InsertIntersectionState(
 
 namespace k2 {
 
-// The implementation of this function is inspired by
+// This function uses "Tarjan's strongly connected components algorithm"
+// (see
+// https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm)
+// to find co-accessible states in a single pass of the graph.
+//
+// The notations "lowlink", "dfnumber", and "onstack" are from the book
+// "The design and analysis of computer algorithms", on page 192,
+// Fig. 5.15 "Procedure to compute LOWLINK", written by John E. Hopcroft.
+//
 // http://www.openfst.org/doxygen/fst/html/connect_8h_source.html
```

Comment:
I am a little concerned by the complexity of this.  If it were me I would have implemented it in about 10 lines of code; that would be less bug-prone and easier to transfer it to GPU.  I'll merge, but I think it's something we should revisit later.

Reply:
The current approach is difficult for parallelization.

I can change the implementation if your pseudo code is available.

Reply:
The current implementation visits every node and every edge only once, so the
time complexity is O(V + E).

We need several `vectors` to save the node state, so the space complexity is 
`some constant * V`